### PR TITLE
Update servers docs from RC1 to RTM

### PR DESCRIPTION
**Pull request by halter73**
@Tratcher @rachelappel

### CODE REVIEW

**Comment by halter73** [comment]
This should also address #1348

File: aspnet/fundamentals/servers.rst [hunk 1]
```diff
@@ -13,39 +13,44 @@ ASP.NET Core is completely decoupled from the web server environment that hosts
 
 `View or download sample code <https://github.com/aspnet/Docs/tree/master/aspnet/fundamentals/servers/sample>`__
 
-Servers and commands
---------------------
+Servers and WebHostBuilderExtensions
+------------------------------------
 
-ASP.NET Core was designed to decouple web applications from the underlying HTTP server. Traditionally, ASP.NET apps have been windows-only hosted on Internet Information Server (IIS). The recommended way to run ASP.NET Core applications on Windows is using IIS as a reverse-proxy server. The HttpPlatformHandler module in IIS manages and proxies requests to an HTTP server hosted out-of-process. ASP.NET Core ships with two different HTTP servers:
+ASP.NET Core was designed to decouple web applications from the underlying HTTP server. Traditionally, ASP.NET apps have been windows-only hosted on Internet Information Server (IIS). The recommended way to run ASP.NET Core applications on Windows is using IIS as a reverse-proxy server. The ASP.NET Core Module in IIS manages and proxies requests to the Kestrel HTTP server hosted out-of-process. ASP.NET Core ships with two different HTTP servers:
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: "The recommended way to run ASP.NET Core applications on Windows is still using IIS, but as a reverse-proxy server."

File: aspnet/fundamentals/servers.rst [hunk 2]
```diff
@@ -13,39 +13,44 @@ ASP.NET Core is completely decoupled from the web server environment that hosts
...
+The default web host for ASP.NET apps developed using Visual Studio is IIS Express functioning as a reverse proxy server for Kestrel. The "Microsoft.AspNetCore.Server.Kestrel" and "Microsoft.AspNetCore.Server.IISIntegration" dependencies are included in *project.json* by default, even with the Empty web site template. Visual Studio provides support for multiple profiles. In addition to the default profile for running in IIS Express, the templates include a second profile that executes the app directly relying on Kestrel for self-hosting. You can manage these profiles and their settings in the **Debug** tab of your web application project's Properties menu or from the *launchSettings.json* file.
 
 .. image:: /fundamentals/servers/_static/serverdemo-properties.png
 
-.. note:: IIS doesn't support commands. Visual Studio launches IIS Express and loads the application with the selected profile.
+.. note:: IIS supports proxying requests to Kestrel but **not** WebListener.
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: AspNetCoreModule

File: aspnet/fundamentals/servers.rst (hunk 1 above)
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: The recommended way to run ASP.NET Core applications on Linux is by using Nginx as a reverse proxy to Kestrel.  Sites may also be run directly using Kestrel or WebListener, but this is only supported for development scenarios at the moment.

File: aspnet/fundamentals/servers.rst [hunk 3]
```diff
@@ -13,39 +13,44 @@ ASP.NET Core is completely decoupled from the web server environment that hosts
...
+
+``dotnet run`` will launch the application from the ``Main`` method. Kestrel is selected by default in the sample project. The sample is programmed so WebListener can be selected instead by passing ``--server WebListener`` as a command line argument. The sample explicitly reads the ``--server`` command line argument in ``Program.Main`` to determine whether to call :dn:method:`~Microsoft.AspNetCore.Hosting.WebHostBuilderKestrelExtensions.UseKestrel` or :dn:method:`~Microsoft.AspNetCore.Hosting.WebHostBuilderWebListenerExtensions.UseWebListener`. The ``--server`` command line flag is **not** interpreted by the ASP.NET Core framework to have any special meaning.
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: "The Main method is the entry point for the application." Remove the part about dotnet run, there are several other common ways to launch the app.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs [hunk 4]
```diff
@@ -1,43 +1,93 @@
...
-                .UseServer("Microsoft.AspNet.Server.Kestrel")
-                .Build()
-                .Start())
+            Server = config["server"] ?? "Kestrel";
+
+            var builder = new WebHostBuilder()
+                .UseContentRoot(Directory.GetCurrentDirectory())
+                .UseStartup<Startup>();
+
+            // The default url is http://localhost:5000 if none is specified.
```
Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs]: doc that `urls` is the normal key, and the ASPNETCORE_URLS env var is read automatically.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs [hunk 5]
```diff
@@ -1,43 +1,93 @@
...
+                Console.WriteLine("IIS cannot be specified at runtime since it does not support self-hosting.");
+                return 1;
+            }
+
+            // ASPNETCORE_PORT is the port that IIS proxies requests to.
+            if (Environment.GetEnvironmentVariable($"ASPNETCORE_PORT") != null)
+            {
+                if (string.Equals(Server, "WebListener", StringComparison.OrdinalIgnoreCase))
+                {
+                    Console.WriteLine("Error: WebListener cannot be used with IIS");
```
Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs]: AspNetCoreModule

File: aspnet/fundamentals/servers.rst [hunk 6]
```diff
@@ -95,55 +100,32 @@ ASP.NET defines a number of :doc:`request-features`. The following table lists t
...
-The *project.json* file shown above demonstrates how to pass the ``server.urls`` parameter directly:
-
-.. code-block:: javascript
-
-  "web": "Microsoft.AspNetCore.Kestrel --server.urls http://localhost:5004"
+Port 0 binding with Kestrel
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Alternately, a  JSON configuration file can be used,
+The :dn:interface:`~Microsoft.AspNetCore.Builder.IApplicationBuilder` available in the ``Startup``'s ``Configure`` method exposes ``ServerFeatures`` of type :dn:interface:`~Microsoft.AspNetCore.Http.Features.IFeatureCollection`. Kestrel and WebListener both expose only a single :dn:interface:`~Microsoft.AspNetCore.Hosting.Server.Features.IServerAddressesFeature`, but different server implementations may expose additional functionality.
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: Move this up, it's not related to Port 0 binding.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs (hunk 4 above)
Comment (halter73) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs]: Where would I put this? How would developers use the info? It seems like an irrelevant implementation detail for most customers.

**Comment by Rick-Anderson** [comment]
@halter73 
c# 6 won't lex, use none in lieu of c#
154/home/travis/build/aspnet/Docs/aspnet/fundamentals/servers.rst:110: WARNING: Could not lex literal_block as "c#". Highlighting skipped.

File: aspnet/fundamentals/servers.rst [hunk 7]
```diff
@@ -152,18 +134,14 @@ WebListener
...
-You can add support for WebListener to your ASP.NET application by adding the "Microsoft.AspNetCore.Server.WebListener" dependency in *project.json* and the following command:
-
-.. code-block:: javascript
-
-  "web": "Microsoft.AspNetCore.Hosting --server Microsoft.AspNetCore.Server.WebListener --server.urls http://localhost:5000"
+You can add support for WebListener to your ASP.NET application by adding the ``Microsoft.AspNetCore.Server.WebListener`` dependency in *project.json* and calling :dn:meth:`~Microsoft.AspNetCore.Hosting.WebHostBuilderWebListenerExtensions.UseWebListener`
 
 .. _kestrel:
 
 Kestrel
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: Move Kestrel above WebListener, WebListener is still prerelease.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs (hunk 4 above)
Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Program.cs]: This sample seems to say developers should define their own command line parameter and call UseUrls. They'd be better off using the existing parameter. If you just want to demonstrate UseUrls don't mix it with Configuration.

File: aspnet/fundamentals/servers.rst (hunk 6 above)
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: I just mean that ServerFeatures should get its own section followed by one for Kestrel.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs [hunk 8]
```diff
@@ -1,55 +1,47 @@
...
-                                        serverAddress);
-                await context.Response.WriteAsync(message);
+                await context.Response.WriteAsync($"Hosted by {Program.Server}\r\n\r\n");
+
+                if (serverAddressesFeature != null)
+                {
+                    await context.Response.WriteAsync($"Listening on the following addresses: {string.Join(", ", serverAddressesFeature.Addresses)}\r\n");
+                }
+
+                await context.Response.WriteAsync($"Local connection info: {context.Connection.LocalIpAddress}:{context.Connection.LocalPort}\r\n");
```
Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs]: Does this show `::1:5000` for IPv6? It might need square brackets or some other formatting.

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs [hunk 9]
```diff
@@ -1,55 +1,47 @@
...
+                await context.Response.WriteAsync($"Remote connection info: {context.Connection.RemoteIpAddress}:{context.Connection.RemotePort}\r\n\r\n");
+
+                var requestUrl = $"{context.Request.Scheme}://{context.Request.Host}{context.Request.PathBase}{context.Request.Path}{context.Request.QueryString}";
```
Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs]: There's a HttpRequest.GetDisplayUrl() extension for this in Http.Extensions.

**Comment by Rick-Anderson** [comment]
@halter73 let me know if you need any help with this.

File: aspnet/fundamentals/servers.rst (hunk 2 above)
Comment (halter73) [inline aspnet/fundamentals/servers.rst]: I went with "The IIS ASP.NET Core Module"

Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: @shirhatti?

Comment (shirhatti) [inline aspnet/fundamentals/servers.rst]: "ASP.NET Core Module" for IIS

File: aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs (hunk 8 above)
Comment (halter73) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs]: Correct. I'm open to any way to write this that's clean enough for sample code. Maybe just remove those two lines?

Comment (Tratcher) [inline aspnet/fundamentals/servers/sample/ServersDemo/src/ServersDemo/Startup.cs]: Yeah, we don't need these two lines in this sample.

File: aspnet/fundamentals/servers.rst [hunk 10]
```diff
@@ -95,77 +100,54 @@ ASP.NET defines a number of :doc:`request-features`. The following table lists t
...
-  "web": "Microsoft.AspNetCore.Hosting --server Microsoft.AspNetCore.Server.WebListener --server.urls http://localhost:5000"
+.. _weblistener:
 
-.. _kestrel:
+WebListener
+-----------
 
-Kestrel
--------
+WebListener is a Windows-only HTTP server for ASP.NET Core. It runs directly on the `Http.Sys kernel driver <http://www.iis.net/learn/get-started/introduction-to-iis/introduction-to-iis-architecture>`_, and has very little overhead.
```
Comment (Tratcher) [inline aspnet/fundamentals/servers.rst]: WebListener cannot be used with the ASP.NET Core Module, it can only be used independently.

**Comment by Tratcher** [comment]
:shipit:

**Comment by Rick-Anderson** [comment]
@halter73  can I merge this now?

**Comment by Rick-Anderson** [comment]
@halter73 do you want to put your name on this as co-author?

**Comment by halter73** [comment]
@Rick-Anderson I squashed/rebased and updated the authors. I think it's good to go. I can push to dev if you want.
